bitswap/client/internal/messagequeue/messagequeue.go

```diff
@@ -476,27 +476,43 @@ func (mq *MessageQueue) rebroadcastWantlist() {
 	mq.rebroadcastIntervalLk.Unlock()
 
 	// If some wants were transferred from the rebroadcast list
-	if mq.transferRebroadcastWants() {
+	if toRebroadcast := mq.transferRebroadcastWants(); toRebroadcast > 0 {
 		// Send them out
 		mq.sendMessage()
+		log.Infow("Rebroadcasting wants", "amount", toRebroadcast, "peer", mq.p)
 	}
 }
 
 // Transfer wants from the rebroadcast lists into the pending lists.
-func (mq *MessageQueue) transferRebroadcastWants() bool {
+func (mq *MessageQueue) transferRebroadcastWants() int {
 	mq.wllock.Lock()
 	defer mq.wllock.Unlock()
 
-	// Check if there are any wants to rebroadcast
-	if mq.bcstWants.sent.Len() == 0 && mq.peerWants.sent.Len() == 0 {
-		return false
+	mq.rebroadcastIntervalLk.Lock()
+	rebroadcastInterval := mq.rebroadcastInterval
+	mq.rebroadcastIntervalLk.Unlock()
+
+	now := time.Now()
+	var toRebroadcast int
+	for _, want := range mq.bcstWants.sent.Entries() {
+		sentAt, ok := mq.bcstWants.sentAt[want.Cid]
+		if ok && now.Sub(sentAt) >= rebroadcastInterval {
+			mq.bcstWants.pending.Add(want.Cid, want.Priority, want.WantType)
+			mq.bcstWants.sent.Remove(want.Cid)
+			toRebroadcast++
+		}
 	}
 
-	// Copy sent wants into pending wants lists
-	mq.bcstWants.pending.Absorb(mq.bcstWants.sent)
-	mq.peerWants.pending.Absorb(mq.peerWants.sent)
+	for _, want := range mq.peerWants.sent.Entries() {
+		sentAt, ok := mq.peerWants.sentAt[want.Cid]
+		if ok && now.Sub(sentAt) >= rebroadcastInterval {
+			mq.peerWants.pending.Add(want.Cid, want.Priority, want.WantType)
+			mq.peerWants.sent.Remove(want.Cid)
+			toRebroadcast++
+		}
+	}
```

Comment:
The previous call to `mq.peerWants.pending.Absorb` also did:
```
// Invalidate the cache up-front to avoid doing any work trying to keep it up-to-date.                                 
w.cached = nil
```
Evan though all the send entries may not be added to pending in the new code, the cache still needs to be cleared. Otherwise, calling `mq.peerWants.pending.Entries()` will not return the newly added entries. Alternatively, the new entries can be to added to the wantlist's `cached` entries and sorted.

I suggest to keep the previous code:
```
mq.bcstWants.pending.Absorb(mq.bcstWants.sent)
mq.peerWants.pending.Absorb(mq.peerWants.sent)
```
And instead modify the `Absorb` function in `boxo/bitswap/client/wantlist/wantlist.go`. WDYT?

Reply:
This doesn't work because `pending *WantList` doesn't have access to `sentAt` map.

Reply:
I can try to make a new method on `recallWantList` struct tho

Reply:
Ok, check the new version

Reply:
On the cache invalidation note. `WantList.Add` also invalidates the cache, so we don't have to it.